src/box/wal.c

```diff
@@ -971,30 +972,16 @@ wal_assign_lsn(struct vclock *vclock_diff, struct vclock *base,
 			 * only capable of writing to local and
 			 * temporary spaces.
 			 */
-			if ((*row)->group_id != GROUP_LOCAL)
+			if ((*row)->group_id != GROUP_LOCAL) {
 				(*row)->replica_id = instance_id;
+				last_glob_row = row;
```

Comment:
This is the question of code readability, it's up to you, to fix it or not. I don't like, that we duplicate assigning `last_glob_row` in two mutually exclusive `if` code blocks. We can do it below, alongside with `first_glob_row`.

<details><summary>Diff</summary>

```diff
diff --git a/src/box/wal.c b/src/box/wal.c
index f28d6211e..dd65e34ed 100644
--- a/src/box/wal.c
+++ b/src/box/wal.c
@@ -974,14 +974,12 @@ wal_assign_lsn(struct vclock *vclock_diff, struct vclock *base,
 			 */
 			if ((*row)->group_id != GROUP_LOCAL) {
 				(*row)->replica_id = instance_id;
-				last_glob_row = row;
 			}
 
 			(*row)->lsn = vclock_inc(vclock_diff, (*row)->replica_id) +
 				      vclock_get(base, (*row)->replica_id);
 		} else {
 			int64_t diff = (*row)->lsn - vclock_get(base, (*row)->replica_id);
-			last_glob_row = row;
 			if (diff <= vclock_get(vclock_diff,
 					       (*row)->replica_id)) {
 				int64_t confirmed_lsn =
@@ -999,16 +997,18 @@ wal_assign_lsn(struct vclock *vclock_diff, struct vclock *base,
 			/* Reset row flags taken from the remote instance. */
 			(*row)->flags = 0;
 		}
+
 		/*
 		 * Use lsn of the first global row as
 		 * transaction id.
 		 */
-		if ((*row)->group_id != GROUP_LOCAL && tsn == 0) {
-			tsn = (*row)->lsn;
-			/*
-			 * Remember the tail being processed.
-			 */
-			first_glob_row = row;
+		if ((*row)->group_id != GROUP_LOCAL) {
+			last_glob_row = row;
+			if (tsn == 0) {
+				tsn = (*row)->lsn;
+				/** Remember the tail being processed. */
+				first_glob_row = row;
+			}
 		}
 		(*row)->tsn = tsn == 0 ? (*start)->lsn : tsn;
 	}

```

</details>

Reply:
It seems I'll have to dump this commit and instead make relay transactional.